database-plugins/src/main/java/io/cdap/plugin/db/connector/DBConnector.java

```diff
@@ -0,0 +1,316 @@
+
+  private static final String RESULTSET_COLUMN_TABLE_NAME = "TABLE_NAME";
+  private static final String RESULTSET_COLUMN_TABLE_TYPE = "TABLE_TYPE";
+  private static final String RESULTSET_COLUMN_TABLE_SCHEM = "TABLE_SCHEM";
+  private static final String RESULTSET_COLUMN_DATA_TYPE = "DATA_TYPE";
+  private static final String RESULTSET_COLUMN_TYPE_NAME = "TYPE_NAME";
+  private static final String RESULTSET_COLUMN_DECIMAL_DIGITS = "DECIMAL_DIGITS";
+  private static final String RESULTSET_COLUMN_COLUMN_SIZE = "COLUMN_SIZE";
+  private static final String RESULTSET_COLUMN_COLUMN_NAME = "COLUMN_NAME";
+  private static final String RESULTSET_COLUMN_IS_NULLABLE = "IS_NULLABLE";
+
+  private DBConnectorConfig config;
+  private DriverCleanup driverCleanup;
+
+  DBConnector(DBConnectorConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configure(PluginConfigurer configurer) {
+    Class<? extends Driver> driverClass = DBUtils.loadJDBCDriverClass(configurer, config.getJdbcPluginName(),
+      String.format("connector.jdbc.%s", config.getJdbcPluginName()), null);
+    try {
+      driverCleanup =
+        DBUtils.ensureJDBCDriverIsAvailable(driverClass, config.getConnectionString(), config.getJdbcPluginName());
+    } catch (Exception e) {
+      throw new RuntimeException(String
+        .format("Failed to register JDBC driver Class, connection string : %s, JDBC plugin name : %s.",
+          config.getConnectionString(), config.getJdbcPluginName()), e);
+    }
+  }
+
+  @Override
+  public void test(FailureCollector collector) throws ValidationException {
+    try {
+      java.sql.Connection connection = getConnection(config.getConnectionString(), config.getAllConnectionArguments());
+      connection.getMetaData();
+    } catch (Exception e) {
+      collector.addFailure(String.format("Failed to connect to the database : %s.", e.getMessage()),
+        "Make sure you " + "specify the correct connection properties.").withStacktrace(e.getStackTrace());
+    }
+  }
+
+  @Override
+  public BrowseDetail browse(BrowseRequest browseRequest) throws IOException {
+    return connectAndQuery(browseRequest.getPath(), "browse", (connection, path) -> {
+      String database = path.getDatabase();
+      if (database == null) {
+        // browse project to list all databases
+        return listDatabases(connection, browseRequest.getLimit());
+      }
+      String schema = path.getSchema();
+      if (schema == null && path.supportSchema()) {
+        return listSchemas(connection, database, browseRequest.getLimit());
+      }
+      String table = path.getTable();
+      if (table == null) {
+        return listTables(connection, database, schema, browseRequest.getLimit());
+      }
+      return getTableDetail(connection, database, schema, table);
+    });
+  }
+
+  @Override
+  public List<StructuredRecord> sample(SampleRequest sampleRequest) throws IOException {
```

Comment:
I feel here it is better to have this class implement `BatchConnector`, where you only need to provide the input format provider, https://github.com/cdapio/hydrator-plugins/blob/develop/database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java#L174, then you can just reuse the util class like this for the record https://github.com/cdapio/hydrator-plugins/blob/develop/database-plugins/src/main/java/io/cdap/plugin/db/batch/source/DBSource.java#L187